text/0000-get-expr-type.md

```diff
@@ -0,0 +1,239 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Given some value `x`, we can retrieve its type with just `x.type`. This type can
+then be used within various contexts:
+
+```rust
+let x = "hellooo";
+let y: x.type = x; // C equivalent: typeof(x) y = x;
+
+type X = x.type;
+
+assert_eq_type!(X, &str); // taken from `static_assertions`
+
+assert!(<x.type>::default().is_empty());
+```
+
+Note that if the expression resolves to a long operation, the operation will
+_not_ be evaluated.
+
+```rust
+type T = do_stuff().do_more().type;
+```
+
+This feature is especially useful within macros when wanting to generate
+functions or types based off of the given expression.
+
+```rust
+let x: Value = /* ... */;
+
+macro_rules! do_stuff {
+    ($x:expr) => {
+        let y = <$x.type>::new(/* ... */);
+        $x.do_stuff_with(y);
+    }
+}
+
+do_stuff!(x);
+```
+
+When we get `x.type` here, it is no different than just substituting it directly
+with `Value`. This allows for accessing all type-specific functionality.
+
+This isn't possible with the current mechanism for getting the type of an
+expression:
+
+```rust
+macro_rules! do_stuff {
+    ($x:expr) => {
+        fn with_type_of<T>(x: T) {
+            let y = T::new(/* ... */);
+            x.do_stuff_with(y);
+        }
+        with_type_of($x);
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+I am uncertain of how this would be implemented within the compiler but I
+imagine that it would leverage the work that's already been done with `const`
+generics.
+
+Some cases to be aware of are non-simple expressions, such as `1 + 1`. It is
+understood that this has a type of `i32` by default, but such an expression may
+be difficult to parse in a generic context.
+
+The following may only be possible in this form:
+
+```rust
+do_stuff::<{1 + 1}.type>();
+```
+
+See [unresolved questions](#unresolved-questions) for more regarding the above
+example.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+By using a postfix syntax, one may write a long expression only to realize that
+nothing before the `.type` part will actually not be evaluated. However, such
+code would be in a context where it's obvious that the expression isn't
+important.
+
+```rust
+do_thing::<send_request().await?.body().await?.contents.type>();
+```
+
+To be fair, this isn't the [weirdest thing in Rust](https://github.com/rust-lang/rust/blob/master/src/test/run-pass/weird-exprs.rs).
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Retrieving the type via the overloaded `.` operator feels like a natural
+extension of the language.
+
+As of this writing, the lang team [has come to a final
+decision](https://boats.gitlab.io/blog/post/await-decision-ii/) regarding the
+syntax for `await`, and that is to make it a postfix operation: `expr.await`.
```

Comment:
The reason `.await` was chosen is because we can chain stuff *after* the `.await`. Similarly, `.match` is an operation which takes an expression and returns an expression, allowing chaining.

There is no such advantage for a type:

* `&'a X`, `*const X` are _prefix_ operators
* `[X]`, `(X, Y)` are _circumfix_
* You cannot write `foo.type::Stuff` since `foo.type` isn't a path, so you'll need `<foo.type>::Stuff` (or drastically change the grammar), making associated type another _circumfix_ operator.

So I don't see any reason why `x.type` is picked instead of `typeof(x)` (or `typeof x` to match `impl Trait`/`dyn Trait`).

Reply:
I do prefer the syntax `<typeof x>::Stuff` over `<x.type>::Stuff` and so I may change the proposal to use this instead. It's less visual noise in my opinion.